### PR TITLE
Create detailed_log.py

### LINKED ISSUE #83: Improve the log summary script
The log summary need to be edited to track the boss you really hunted for (otherwise, it displays useless lines). We could avoid this. See PR #82 for more informations

### PR DESCRIPTION
**Pull request by TeachingGame**
Detiled log file that returns non legend shiny names and global legend odds.

### CODE REVIEW

**Review by pifopi — APPROVED** [review]
Well done ! I let other people decide if this should go in the older file or not. I'm fine with both !

**Comment by denvoros** [comment]
I like this contribution quite a bit! Thanks for submitting.

My one concern running it is that it spits out information about *every single legendary* even if you didn't run the Pokémon. I know that there's a comment that says to edit the list of bosses for what you want, but I don't think that having the user edit the long list of strings is very friendly and intuitive, especially considering many users of this script are not going to be comfortable editing the script.

I'd recommend instead running a check before the print statement to check if any information was found. This will require a little bit of rework since you also have loop logic that's fairly inefficient.

Instead of looping over the bosses and then looping over all of the files, why not find the files that belong to the boss before-hand? Make a `dict` object, and in a separate loop before your current boss loop find all files that contain the boss name and `.txt` and `.log` in a list and map that to the dictionary. If there are no files, you can ignore adding the boss. You could do something like this:

```python
boss_files = {}
for boss in bosses:
    found_files = []
    for x in file_list:
        if boss.lower() in x.lower() and ('.log' in x or '.txt' in x):
            # add to the files found
            found_files.append(x)
            # remove from the original list since we're done with it
            file_list.remove(x)

    if len(found_files) > 0:
        boss_files[boss] = found_files
```

Then, if there are no files for the boss, you just don't print the information about that boss. It reduces your loop iteration over the files while also allowing for uses to not edit the long list of strings to get global stats. If you follow the loop above, then your "search loop" becomes:

```python
for boss, curr_file_list in boss_files.items():
    # boss is now the boss string, like before
    # curr_file_list is now a list of the files that match the boss itself, and can be used in iteration/reading
```

As it stands, I'm fine with this being merged into beta, but I think suggestions like I've included will strengthen the utility of the script significantly. I'd be happy to help with my suggestions, and thanks so much for the contribution!

**Review by denvoros — APPROVED** [review]

**Review by ercdndrs — APPROVED** [review]

**Comment by TeachingGame** [comment]
I'll attempt making the changes to include a dictionary and reduce unnecessary boss info this weekend unless it has already been done.

**Review by ercdndrs — CHANGES_REQUESTED** [review]
Changing my feedback to "Request changes" to indicate that further changes are planned.

**Comment by TeachingGame** [comment]
I have got it running with the new system after hours of research and experimentation. Now for some reason this code is not reading all the logs. It seems to miss a couple in the initial collection stage, for example it found only 6 of my 9 latias logs. The missing logs seem to be have identical naming conventions as the found ones.
If you would like to try and figure it out just add
`            print (x)`
to line 34 to see all collected logs
Also, I am unsure how to fix the linting error.

File: scripts/detailed_log.py [hunk 1]
```diff
@@ -8,35 +8,42 @@
...
+boss_files = {}
 
 for boss in bosses:
+    found_files = []
+    for x in file_list:
+        if boss.lower() in x.lower() and ('.log' in x or '.txt' in x):
+            # add to the files found
+            found_files.append(x)
+            # remove from the original list since we're done with it
+            file_list.remove(x)
```
Comment (pifopi) [inline scripts/detailed_log.py]: Your error is due to this. I'm gonna check but usually removing objects from a list while iterating over it leads to issues. You don't need to remove it from the list anyway

File: scripts/detailed_log.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import os
+import sys
```
Comment (pifopi) [inline scripts/detailed_log.py]:
```suggestion
```

File: scripts/detailed_log.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import re
```
Comment (pifopi) [inline scripts/detailed_log.py]:
```suggestion
```

File: scripts/detailed_log.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+# We need to import log files from the parent directory.
+from os.path import dirname, abspath, join
```
Comment (pifopi) [inline scripts/detailed_log.py]:
```suggestion
from os.path import dirname, abspath
```

File: scripts/detailed_log.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+for boss, logs in boss_files.items():
+    total_losses = 0
+    total_wins = 0
+    total_shinies = 0
+    shiny_names = []
+    total_legends = 0
+
+    for log in logs:
+        with open(os.path.join(log_path, log), newline='\n', encoding='utf-8') as log_file:
+                num_losses = 0
```
Comment (pifopi) [inline scripts/detailed_log.py]:
```suggestion
            num_losses = 0
```

Comment (pifopi) [reply scripts/detailed_log.py]: and do like that for all following lines (remove one indentation level)

File: scripts/detailed_log.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    global_wins += total_wins
+    global_losses += total_losses
+    global_shinies += total_shinies
+    global_legends += total_legends
+
+global_runs = global_losses + global_wins
+global_win_percentage = 0 if global_runs == 0 else global_wins / global_runs * 100
+global_legend_odds = 0 if global_legends == 0 else global_wins / global_legends
+
+print(f'\nGlobal Summary')
```
Comment (pifopi) [inline scripts/detailed_log.py]:
```suggestion
print('\nGlobal Summary')
```

**Review by pifopi — CHANGES_REQUESTED** [review]

**Comment by TeachingGame** [comment]
Thanks for the suggestions, I manually adjusted my file on laptop, 1 thing at a time running it to check it worked, then replaced the file in my fork. But I notice now a button next to all your comments, Pif, `commit suggestion.` I'm guessing this would have automatically done the changes you suggested?
It passed the linting error and it now finds the logs it skipped previously.

**Review by pifopi — APPROVED** [review]
You're welcome! Thanks for the change, I vote up !

Yes it would have done just that, but doing it manually is the same ;)

**Comment by TeachingGame** [comment]
One last edit. This file requires no user edits,etc before running.
I added `input('Press ENTER to exit log...')` to the final line so that it can now be ran by doubled clicking the file.

**Review by ercdndrs — APPROVED** [review]

**Pull request merged**